### PR TITLE
Optimizer closure

### LINKED ISSUE #2785: Pass second-order closure to all optimizers (not just LBFGS)
## 🚀 Feature

I could be wrong, but I noticed the following in the code of lightning module's optimizer_step

```python
        if on_tpu:
            xm.optimizer_step(optimizer)
        elif using_native_amp:
            self.trainer.scaler.step(optimizer)
        elif using_lbfgs:
            optimizer.step(second_order_closure)
        else:
            optimizer.step()

```

If someone uses a custom optimizer that needs the closure returning the loss multiple times, it won't work. 

### Pitch

Since all classes that inherit from `torch.optim.Optimizer` have the step method accept a closure (even if they don't need it), we could just do

```python
        if on_tpu:
            xm.optimizer_step(optimizer)
        elif using_native_amp:
            self.trainer.scaler.step(optimizer)
        # elif using_lbfgs:
        #     optimizer.step(second_order_closure)
        else:
            optimizer.step(second_order_closure)
```
and drop the "using_lbfgs" argument?

### Alternatives

The user has to override the optimizer_step themself.

### LINKED ISSUE #4188: To many backwards with LBFGS
## 🐛 Bug

When using LBFGS we have one backward step to much, because we call backward before the optimiser step (also for gradient accumulation), but the optimizer step get's a closure and therefore calls backward again.

### To Reproduce
```
import torch
import pytorch_lightning as ptl
from pytorch_lightning import LightningModule
from torch.utils.data import Dataset


class RandomDictDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        a = self.data[index]
        b = a + 2
        return {"a": a, "b": b}

    def __len__(self):
        return self.len


class RandomDictStringDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return {"id": str(index), "x": self.data[index]}

    def __len__(self):
        return self.len


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        """
        Testing PL Module
        Use as follows:
        - subclass
        - modify the behavior for what you want
        class TestModel(BaseTestModel):
            def training_step(...):
                # do your own thing
        or:
        model = BaseTestModel()
        model.training_epoch_end = None
        """
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.cross_entropy(
            prediction,
            torch.ones(len(prediction), dtype=torch.long, device=prediction.device),
        )

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        self.log("loss", loss)
        return loss

    def validation_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        self.log("loss", loss)
        return loss

    def test_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.LBFGS(self.parameters())
        return optimizer

    def train_dataloader(self):
        return torch.utils.data.DataLoader(RandomDataset(32, 64), batch_size=16)

    def val_dataloader(self):
        return torch.utils.data.DataLoader(RandomDataset(32, 64), batch_size=16)

    def test_dataloader(self):
        return torch.utils.data.DataLoader(RandomDataset(32, 64), batch_size=16)


def main():
    model = BoringModel()
    trainer = ptl.Trainer(
        distributed_backend="dp",
        gpus=1,
    )
    trainer.fit(model)


if __name__ == "__main__":
    main()```

### Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

* CUDA:
        - GPU:
                - GeForce RTX 2080 Ti
                - GeForce RTX 2080 Ti
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.19.2
        - pyTorch_debug:     False
        - pyTorch_version:   1.6.0
        - pytorch-lightning: 20201015
        - tqdm:              4.50.2
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.5
        - version:           #52~18.04.1-Ubuntu SMP PREEMPT Thu Sep 10 13:34:23 UTC 2020

### Additional context

Probably we can fix this. by passing a closure to all optimisers (to be more consistent).

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Fixes #4188 and also fixes #2785 

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
Fixed https://github.com/PyTorchLightning/pytorch-lightning/issues/4083 too right?

**Comment by justusschock** [comment]
@rohitgr7 Not yet tested, net to get it working first :D

**Comment by williamFalcon** [comment]
nice. can you add tests that use the mock + count to verify that we perform the expected number of backward passes? also with a non-lbfgs optimizer to make sure we don't break current functionality

**Comment by justusschock** [comment]
@williamFalcon I think @awaelchli is currently looking into this.

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1202,13 +1200,16 @@ def optimizer_step(self, current_epoch, batch_idx, optimizer,
...
+            xm.optimizer_step(optimizer, {'closure': optimizer_closure})
         elif using_native_amp:
+            # native amp does not yet support closures.
+            # TODO: pass the closure to the step ASAP
+            optimizer_closure()
             self.trainer.scaler.step(optimizer)
-        elif using_lbfgs:
-            optimizer.step(second_order_closure)
+        # elif using_lbfgs:
+        #     optimizer.step(second_order_closure)
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1210]:
```suggestion
```

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -667,70 +670,73 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
...
+
+                        # optimizer step
+                        self.optimizer_step(optimizer, opt_idx, batch_idx, train_step_and_backward_closure)
+
+                    else:
+                        self._curr_step_result = self.training_step(split_batch, batch_idx, opt_idx, self.trainer.hiddens)
+
+                    # TODO: we may also have to check current_result.loss if it is None
+                    # this can happen when manual optimization does not return a loss but
+                    # dict/result for logging
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:714]:
```suggestion
```
my comment here  is outdated, and is covered by the manual_optimization tests

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -1202,13 +1200,14 @@ def optimizer_step(self, current_epoch, batch_idx, optimizer,
 
         """
         if on_tpu:
-            xm.optimizer_step(optimizer)
+            xm.optimizer_step(optimizer, {'closure': optimizer_closure})
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1203]: we should check what's going on here, this is so far the only obvious place I found where we touch tpu logic
cc @lezwon

Comment (lezwon) [reply pytorch_lightning/core/lightning.py:1203]: Yeah, i'm not sure it matches the function signature of `optimizer_step`. https://github.com/pytorch/xla/blob/f572937c4da8f382b1aa8448b23563df0084ec4f/torch_xla/core/xla_model.py#L755

Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1203]:
```suggestion
            xm.optimizer_step(optimizer, optimizer_args={'closure': optimizer_closure})
```

Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1203]: thanks @lezwon 
I think this is the fix then?

**Review by awaelchli — APPROVED** [review]
🚀

**Review by tchaton — APPROVED** [review]
Great addition !

**Review by SeanNaren — APPROVED** [review]

**Comment by awaelchli** [comment]
The GPU tests did not run, because of horovod installation failing, so we did not see the problem with amp closure.
